### PR TITLE
[GEOT-5859] upgrade httpclient

### PR DESCRIPTION
**Pull request by ianturton**
[GEOT-5859](https://osgeo-org.atlassian.net/browse/GEOT-5859) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geotools&utm_campaign=badge_info)

<Include a few sentences describing the overall goals for this Pull Request>

# Checklist

Reviewing is a process done by project maintainers, **mostly on a volunteer basis** (thus limited in time). We need to keep the review overhead as small as possible, and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geotools/geotools/blob/main/CONTRIBUTING.md)
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the `main` branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.
- [x] The changes are not causing two modules to share the same Java packages (to avoid [Java 9+ split package](http://tutorials.jenkov.com/java/modules.html#split-packages-not-allowed) issues)
- [x] The changes are not breaking the build in downstream projects using SNAPSHOT dependencies, GeoWebCache and GeoServer (there is an automatic PR check verifying this, check this when it turns green).

The following are required only for core and extension modules (they are welcomed, but not required, for unsupported modules):

- [x] There is an issue in [Jira](https://osgeo-org.atlassian.net/projects/GEOT) describing the bug/task/new feature (a notable exemptions is, changes not visible to end users). The ticket is for the GeoTools project, if the issue was found elsewhere it's a good practice to link to the origin ticket/issue.
- [x] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message(s) must be in the form "[GEOT-XYZW] Title of the Jira ticket"
- [x] New unit tests have been added covering the changes
- [x] This PR passes all existing unit tests (test results will be reported by Continuous Integration after opening this PR)
- [x] This PR passes the [QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html) (QA checks results will be reported by Continuous Integration after opening this PR)
- [ ] Documentation has been updated accordingly.

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or not applicable.

### CODE REVIEW

**Comment by ianturton** [comment]
It would be good @roarbra had a look at this too - I think that all the local/nonlocal proxy stuff is handled by the HttpClient code now, but I'm not sure how to test it to make sure.

**Review by mprins — APPROVED** [review]
Great, LGTM!
I seem to recall this as a lot of work having logged the initial issue, but perhaps the refactoring alleviated that somewhat.

**Review by mprins — APPROVED** [review]

File: modules/plugin/http-commons/src/main/java/org/geotools/http/commons/MultithreadedHttpClient.java [hunk 1]
```diff
@@ -77,123 +78,83 @@
...
-            HttpState state = client.getState();
-            if (state == null) {
-                state = new HttpState();
-                client.setState(state);
-            }
-            AuthScope authscope = AuthScope.ANY;
-            Credentials credentials = new UsernamePasswordCredentials(proxyUser, proxyPassword);
-            state.setProxyCredentials(authscope, credentials);
-        }
+        clientBuilder.useSystemProperties();
```
Comment (aaime) [inline modules/plugin/http-commons/src/main/java/org/geotools/http/commons/MultithreadedHttpClient.java:106]: Oh neat... and it accounts for even more system properties than the old code did, see the [class javadoc](https://hc.apache.org/httpcomponents-client-4.5.x/current/httpclient/apidocs/index.html?org/apache/http/impl/client/HttpClientBuilder.html) (it's not in the method javadoc for some reason).

File: modules/plugin/http-commons/src/main/java/org/geotools/http/commons/MultithreadedHttpClient.java [hunk 2]
```diff
@@ -250,59 +222,68 @@ public void setPassword(String password) {
...
-        client.getState().clearCredentials();
         if (user != null && password != null) {
             AuthScope authscope = AuthScope.ANY;
             Credentials credentials = new UsernamePasswordCredentials(user, password);
-
-            client.getParams().setAuthenticationPreemptive(true);
-            client.getState().setCredentials(authscope, credentials);
-        } else {
-            client.getParams().setAuthenticationPreemptive(false);
+            // TODO - check if this works for all types of auth or do we need to look it up?
```
Comment (aaime) [inline modules/plugin/http-commons/src/main/java/org/geotools/http/commons/MultithreadedHttpClient.java:228]: I see another bit of code above where credentials are base64 encoded and set in the headers... so is this not working as it should?

**Review by aaime — APPROVED** [review]
Overall looks pretty good. I have a question below.

**Comment by aaime** [comment]
@roarbra any chance you could have a look at this one? Upgrades the multithreaded HTTP client to Commons HTTP client 4.5.x

File: modules/plugin/http-commons/src/main/java/org/geotools/http/commons/MultithreadedHttpClient.java (hunk 2 above)
Comment (ianturton) [reply modules/plugin/http-commons/src/main/java/org/geotools/http/commons/MultithreadedHttpClient.java:228]: It seems to depend on if you are using GET or POST and I'm not entirely sure what setting the credentials here does or doesn't affect. Also not sure if this code can handle anything but BASIC auth (as that's all that is tested) or if in fact we support anything else

File: modules/plugin/http-commons/src/main/java/org/geotools/http/commons/MultithreadedHttpClient.java (hunk 1 above)
Comment (ianturton) [reply modules/plugin/http-commons/src/main/java/org/geotools/http/commons/MultithreadedHttpClient.java:106]: Actually, we should probably add that list to our docs somewhere

**Comment by roarbra** [comment]
Hi,

I’m at vacation for the moment, and haven’t a opportunity to look at this
for the next two weeks.

Best regards,

Roar Brænden

man. 12. jul. 2021 kl. 15:43 skrev Ian Turton ***@***.***>:

File: modules/plugin/http-commons/src/main/java/org/geotools/http/commons/MultithreadedHttpClient.java (hunk 2 above)
Comment (ianturton) [reply modules/plugin/http-commons/src/main/java/org/geotools/http/commons/MultithreadedHttpClient.java:240]: I've managed to fix this up - it seems the problem was the wiremock doesn't ask for basic authentication so the client doesn't send it!

**Comment by jodygarnett** [comment]
Writing up release notes, cannot tell what version of http client is actually updated to - is it a transitive dependency now?

**Comment by mprins** [comment]
Commons-httpclient 3.1 upgraded to apache httpclient 4.5.13 as far as I can tell

**Pull request merged**